runtime/src/main/java/org/corfudb/protocols/wireprotocol/IMetadata.java

```diff
@@ -78,10 +79,30 @@ default Long getRank() {
      * @param rank The rank of this append.
      */
     default void setRank(long rank) {
+        if (rank>=0 && getRank()<0) {
+            throw new IllegalStateException("Cannot set rank when the overwrite is permitted");
```

Comment:
Thanks for the detailed comments, suggestions appreciated&accepted, of course.
Next steps:
1) I just resolved the conflicts on the top of your "compact" change and I will update the request
2) I will plug into your compact and unit test, that compact for overwritten entries also works
3) I will change to your approach with Rank + one very carefull change in the locking scope (your existing lock in segment handler), to preserve the atomicity when decide whether to throw an OverwriteException based on rank.
I will keep the changes in this review if I can.